### PR TITLE
Add in Partition type check support

### LINKED ISSUE #1915: [FEA] follow on work for hash partitioning changes
#1910 made hash partitioning match exactly what Spark does.  There were a few issues found and both #1913 and #1914 were filed to cover those.  This is intended to cover some of the other things that need to be done.

- [x] Identity hash for hash_partition and performance checks for it too.
The hash_partition JNI bindings in cudf did not include any way to pass down the hash function to use.  It really would be nice to be able to use hash_partition along with the identity hash function instead of sorting the data and doing upper bounds on it ourselves.  But we need to make sure that the performance does not get worse. I don't see how it could, but we do need to check it.  Be aware that hash_partition is using a regular modulo instead of a pmod so we will still have to calculate the exact partition we want and pass that in instead of just the murmur3 hash function. 

- [x] BitCast updates for Murmur3.
The Murmur3 implementation needed a little bit of work before we could get the individual parts passed in to match what spark was doing.  There are a very few number of places where we need to pull out the raw bits from a value (float/double #1914) and decimal when it is a decimal64 value.  In those cases it really would be nice to write the code in such a way that we can do it without having to materialize a new temporary column.  The float/double is less of a big deal because we have to normalize the NaN values ahead of time, so we will create a temp value anyways.  And for decimal we have to be extra careful because after #1717 it will only work for decimal64.  Decimal32 would still need to have the value copied out.

- [x] performance testing/tuning
Because this was put in as a really fast bug fix at the end of the 0.4 release there wasn't a lot of time to really look at how to performance tune this, or if it is needed.  We should at least do some profiling and see how it performs compared to the previous code.  If we use the identity hash function then really the only extra cost would be computing the murmur3 hash, and the savings would be cudf not computing their murmur3 hash.

- [ ] Type Checking for Partitioning
The current GpuOverrides does not do any type checking for partitioning.  The HashPartitioning does not rewrite the expressions to insert in the murmur3 with the pmod at that time, so we cannot rely on it to do the type checks for us.  The code that went into 0.4 just did the check, but no documentation was generated for it or anything like that.  We should update the type checking and documentation generation to do the right thing.

### PR DESCRIPTION
**Pull request by revans2**
This fixes #1915

I could add in support for input/output checks for Parquet, Orc, and CSV if people want me to, but I thought I would keep it simple to start out with.

This also addresses part of #1755 by outputting headers periodically in the tables. It is better than it was before, but I don't think it is a final solution.

### CODE REVIEW

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 1]
```diff
@@ -601,6 +600,57 @@ object ExecChecks {
...
+  override def tag(meta: RapidsMeta[_, _, _]): Unit = {
+    val part = meta.wrapped
+    val children = meta.childExprs.map(_.wrapped.asInstanceOf[Expression]).toArray
+
+    val fixedChecks = paramCheck.toArray
+    assert (fixedChecks.length <= children.length,
+      s"${part.getClass.getSimpleName} expected at least ${fixedChecks.length} but " +
+          s"found ${children.length}")
+    fixedChecks.indices.foreach { i =>
+      val check = fixedChecks(i)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:622]: Nit: can simplify with something like this:
```suggestion
    fixedChecks.zipWithIndex.foreach { case (check, i) =>
```

**Review by jlowe — DISMISSED** [review]
Only a minor simplification suggestion, lgtm.

**Review by jlowe — APPROVED** [review]

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

**Pull request merged**